pkg/controller/kubelet-config/helpers.go

```diff
@@ -97,8 +99,55 @@ func findKubeletConfig(mc *mcfgv1.MachineConfig) (*ign3types.File, error) {
 	return nil, fmt.Errorf("Could not find Kubelet Config")
 }
 
-func getManagedKubeletConfigKey(pool *mcfgv1.MachineConfigPool, client mcfgclientset.Interface) (string, error) {
-	return ctrlcommon.GetManagedKey(pool, client, "99", "kubelet", getManagedKubeletConfigKeyDeprecated(pool))
+// nolint: dupl
+func getManagedKubeletConfigKey(pool *mcfgv1.MachineConfigPool, client mcfgclientset.Interface, cfg *mcfgv1.KubeletConfig) (string, error) {
+	// Get all the kubelet config CRs
+	kcList, err := client.MachineconfigurationV1().KubeletConfigs().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return "", fmt.Errorf("error listing kubelet configs: %v", err)
+	}
+	// If there is no kubelet config in the list, return the default MC name with no suffix
+	if kcList == nil || kcList.Items == nil {
+		return ctrlcommon.GetManagedKey(pool, client, "99", "kubelet", getManagedKubeletConfigKeyDeprecated(pool))
+	}
+	for _, kc := range kcList.Items {
+		if kc.Name == cfg.Name {
+			val, ok := kc.GetAnnotations()[ctrlcommon.MCNameSuffixAnnotationKey]
+			// If we find a matching kubelet config and it is the only one in the list, then return the default MC name with no suffix
+			if !ok && len(kcList.Items) < 2 {
+				return ctrlcommon.GetManagedKey(pool, client, "99", "kubelet", getManagedKubeletConfigKeyDeprecated(pool))
+			}
+			// Otherwise if an MC name suffix exists, append it to the default MC name and return that as this kubelet config exists and
+			// we are probably doing an update action on it
+			if val != "" {
+				return fmt.Sprintf("99-%s-generated-kubelet-%s", pool.Name, val), nil
+			}
+		}
+	}
+
+	// If we are here, this means that a new kubelet config was created, so we have to calculate the suffix value for its MC name
+	suffixNum := 0
+	// Get the suffix value of the second to last item in the kubelet config list. We use the second to last item because
+	// the most recent kubelet config being created has already been added to the list and is the kubelet config object that we are
+	// trying to figure out the suffix value for
+	val, ok := kcList.Items[len(kcList.Items)-2].GetAnnotations()[ctrlcommon.MCNameSuffixAnnotationKey]
```

Comment:
Are we guaranteed to have 2 items in the slice? len(kcList.Items) could equal 0 - 2, thus an index of -2 or -1.

We need some sort of validation logic before this line.

Reply:
I can add a check, but we are guaranteed to have at least 2 elements in the slice by the time we get to this point. When we create a CR for the first time, https://github.com/openshift/machine-config-operator/pull/2366/files#diff-4a2d0a4442e10d889a927322ad7311c67b22b259622067b52cbda7d2ece3ebefR119 will be true and it will return the original key. And as soon as you create a new CR it is appended to the `Items` list, so while we are figuring out what suffix to use for the CR, it already exists in the `Items` list and that is why the logic `len - 2` so we can get what the suffix was for the previous CR and calculate the suffix for the current one accordingly.